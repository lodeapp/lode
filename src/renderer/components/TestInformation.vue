<template>
    <div class="test-information markdown-body">
        <div v-if="isEmpty">Run the test at least once to gather information.</div>
        <table v-else>
            <colgroup>
                <col width="1%">
                <col width="99%">
            </colgroup>
            <tbody>
                <tr>
                    <td class="heading">Status</td>
                    <td>{{ label }}</td>
                </tr>
                <tr v-if="typeof stats.first !== 'undefined'">
                    <td class="heading">First seen</td>
                    <td :key="timeKey" :title="firstSeen">{{ displayFirstSeen() }}</td>
                </tr>
                <tr>
                    <td class="heading">Last run</td>
                    <td :key="timeKey" v-if="typeof stats.last !== 'undefined'" :title="lastRun">{{ displayLastRun() }}</td>
                    <td v-else>Never</td>
                </tr>
                <tr v-if="typeof stats.duration !== 'undefined'">
                    <td class="heading">Duration</td>
                    <td><Duration :ms="stats.duration" /></td>
                </tr>
                <tr v-if="typeof stats.assertions !== 'undefined'">
                    <td class="heading">Assertions</td>
                    <!-- @TODO: Number formatting -->
                    <td>{{ stats.assertions }}</td>
                </tr>
            </tbody>
        </table>
    </div>
</template>

<script>
import moment from 'moment'
import _isEmpty from 'lodash/isEmpty'
import { labels } from '@lib/frameworks/status'
import Duration from '@/components/Duration'

export default {
    name: 'TestInformation',
    components: {
        Duration
    },
    props: {
        status: {
            type: String,
            default: 'idle'
        },
        stats: {
            type: Object,
            required: true
        }
    },
    data () {
        return {
            label: labels[this.status],
            timeKey: this.$string.random(),
            timeInterval: null
        }
    },
    computed: {
        isEmpty () {
            return _isEmpty(this.stats)
        },
        lastRun () {
            return moment(this.stats.last).format('MMMM Do YYYY, HH:mm:ss')
        },
        firstSeen () {
            return moment(this.stats.first).format('MMMM Do YYYY, HH:mm:ss')
        }
    },
    mounted () {
        this.timeInterval = window.setInterval(() => {
            this.timeKey = this.$string.random()
        }, 15000)
    },
    beforeDestroy () {
        window.clearInterval(this.timeInterval)
    },
    methods: {
        displayLastRun () {
            return moment(this.stats.last).fromNow()
        },
        displayFirstSeen () {
            return moment(this.stats.first).fromNow()
        }
    }
}
</script>
