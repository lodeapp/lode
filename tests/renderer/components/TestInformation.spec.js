import { createLocalVue, shallowMount } from '@vue/test-utils'
import TestInformation from '@/components/TestInformation'
import Strings from '@/plugins/strings'

const localVue = createLocalVue()
localVue.use(new Strings())

const RealDate = Date.now
beforeAll(() => {
    global.Date.now = jest.fn(() => new Date('2020-12-01T14:49:00').getTime())
})
afterAll(() => {
    global.Date.now = RealDate
})

describe('components/TestInformation.vue', () => {
    test.each([
        ['never run', {
            first: '1985-10-21T09:00:00'
        }],
        ['dates only', {
            first: '1985-10-21T09:00:00',
            last: '2015-10-21T13:00:00'
        }],
        ['with duration', {
            first: '2020-09-10T13:00:00',
            last: '2020-11-21T13:00:00',
            duration: 1000
        }],
        ['with long duration', {
            first: '2020-12-01T10:20:00',
            last: '2020-12-01T14:47:00',
            duration: 32198
        }],
        ['with assertions', {
            first: '2020-11-21T13:00:00',
            last: '2020-12-01T14:48:33',
            assertions: 42
        }]
    ])('matches snapshot for stats: "%s"', async (name, stats) => {
        const wrapper = shallowMount(TestInformation, {
            localVue,
            propsData: {
                stats
            }
        })
        expect(wrapper.html()).toMatchSnapshot()
    })
})
