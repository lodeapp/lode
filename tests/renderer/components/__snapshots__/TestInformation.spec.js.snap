// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TestInformation.vue matches snapshot for stats: "dates only" 1`] = `
"<div class=\\"test-information markdown-body\\">
  <table>
    <colgroup>
      <col width=\\"1%\\">
      <col width=\\"99%\\">
    </colgroup>
    <tbody>
      <tr>
        <td class=\\"heading\\">Status</td>
        <td>Idle</td>
      </tr>
      <tr>
        <td class=\\"heading\\">First seen</td>
        <td title=\\"October 21st 1985, 09:00:00\\">about 35 years ago</td>
      </tr>
      <tr>
        <td class=\\"heading\\">Last run</td>
        <td title=\\"October 21st 2015, 13:00:00\\">about 5 years ago</td>
      </tr>
      <!---->
      <!---->
    </tbody>
  </table>
</div>"
`;

exports[`components/TestInformation.vue matches snapshot for stats: "never run" 1`] = `
"<div class=\\"test-information markdown-body\\">
  <table>
    <colgroup>
      <col width=\\"1%\\">
      <col width=\\"99%\\">
    </colgroup>
    <tbody>
      <tr>
        <td class=\\"heading\\">Status</td>
        <td>Idle</td>
      </tr>
      <tr>
        <td class=\\"heading\\">First seen</td>
        <td title=\\"October 21st 1985, 09:00:00\\">about 35 years ago</td>
      </tr>
      <tr>
        <td class=\\"heading\\">Last run</td>
        <td>Never</td>
      </tr>
      <!---->
      <!---->
    </tbody>
  </table>
</div>"
`;

exports[`components/TestInformation.vue matches snapshot for stats: "with assertions" 1`] = `
"<div class=\\"test-information markdown-body\\">
  <table>
    <colgroup>
      <col width=\\"1%\\">
      <col width=\\"99%\\">
    </colgroup>
    <tbody>
      <tr>
        <td class=\\"heading\\">Status</td>
        <td>Idle</td>
      </tr>
      <tr>
        <td class=\\"heading\\">First seen</td>
        <td title=\\"November 21st 2020, 13:00:00\\">10 days ago</td>
      </tr>
      <tr>
        <td class=\\"heading\\">Last run</td>
        <td title=\\"December 1st 2020, 14:48:33\\">less than a minute ago</td>
      </tr>
      <!---->
      <tr>
        <td class=\\"heading\\">Assertions</td>
        <td>42</td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`components/TestInformation.vue matches snapshot for stats: "with duration" 1`] = `
"<div class=\\"test-information markdown-body\\">
  <table>
    <colgroup>
      <col width=\\"1%\\">
      <col width=\\"99%\\">
    </colgroup>
    <tbody>
      <tr>
        <td class=\\"heading\\">Status</td>
        <td>Idle</td>
      </tr>
      <tr>
        <td class=\\"heading\\">First seen</td>
        <td title=\\"September 10th 2020, 13:00:00\\">3 months ago</td>
      </tr>
      <tr>
        <td class=\\"heading\\">Last run</td>
        <td title=\\"November 21st 2020, 13:00:00\\">10 days ago</td>
      </tr>
      <tr>
        <td class=\\"heading\\">Duration</td>
        <td>
          <duration-stub ms=\\"1000\\"></duration-stub>
        </td>
      </tr>
      <!---->
    </tbody>
  </table>
</div>"
`;

exports[`components/TestInformation.vue matches snapshot for stats: "with long duration" 1`] = `
"<div class=\\"test-information markdown-body\\">
  <table>
    <colgroup>
      <col width=\\"1%\\">
      <col width=\\"99%\\">
    </colgroup>
    <tbody>
      <tr>
        <td class=\\"heading\\">Status</td>
        <td>Idle</td>
      </tr>
      <tr>
        <td class=\\"heading\\">First seen</td>
        <td title=\\"December 1st 2020, 10:20:00\\">about 4 hours ago</td>
      </tr>
      <tr>
        <td class=\\"heading\\">Last run</td>
        <td title=\\"December 1st 2020, 14:47:00\\">2 minutes ago</td>
      </tr>
      <tr>
        <td class=\\"heading\\">Duration</td>
        <td>
          <duration-stub ms=\\"32198\\"></duration-stub>
        </td>
      </tr>
      <!---->
    </tbody>
  </table>
</div>"
`;
